src/realm/object-store/sync/app.hpp

```diff
@@ -87,22 +93,18 @@ class App : public std::enable_shared_from_this<App>,
         DeviceInfo device_info;
     };
 
-    // `enable_shared_from_this` is unsafe with public constructors; use `get_shared_app` instead
-    App(const Config& config);
-    App(App&&) noexcept = default;
-    App& operator=(App&&) noexcept = default;
+    // `enable_shared_from_this` is unsafe with public constructors;
+    // use `App::get_app()` instead
+    explicit App(PrivateConstructionOnly, const Config& config);
```

Comment:
nit: you can look at the `MigrationStore` class for a different way to handle this (also inspired from other places)

Reply:
The way I understand it, this is a "graceful" way to prevent direct public construction while also supporting `std::make_shared<App>()`?

Reply:
Yeah, since the type is private nothing outside the class can write the type name, but the class can still use it as a template parameter for other things. This means that wrappers like make_shared can pass along the tag but the tag can't be constructed outside the class.

The advantage it has over the private subclass calling a protected constructor approach is that it doesn't result in a second instantiation of std::shared_ptr for the subclass so it's a little better for compile times (and potentially final binary size if the compiler is insufficiently smart).